webapi/src/main/scala/org/knora/webapi/util/date/CalendarDateUtilV2.scala

```diff
@@ -0,0 +1,543 @@
+            case _ => errorFun
+        }
+    }
+}
+
+/**
+  * Indicates that a date has year precision.
+  */
+case object DatePrecisionYear extends DatePrecisionV2 {
+    override def toString: String = StringFormatter.PrecisionYear
+}
+
+/**
+  * Indicates that a date has month precision.
+  */
+case object DatePrecisionMonth extends DatePrecisionV2 {
+    override def toString: String = StringFormatter.PrecisionMonth
+}
+
+/**
+  * Indicates that a date has day precision.
+  */
+case object DatePrecisionDay extends DatePrecisionV2 {
+    override def toString: String = StringFormatter.PrecisionDay
+}
+
+/**
+  * Represents the name of a calendar.
+  */
+sealed trait CalendarNameV2
+
+object CalendarNameV2 {
+    /**
+      * Parses the name of a calendar.
+      *
+      * @param calendarNameStr a string representing the name of the calendar.
+      * @param errorFun a function that throws an exception. It will be called if the string cannot
+      *                 be parsed.
+      * @return a [[CalendarNameV2]] representing the name of the calendar.
+      */
+    def parse(calendarNameStr: String, errorFun: => Nothing): CalendarNameV2 = {
+        calendarNameStr match {
+            case StringFormatter.CalendarGregorian => CalendarNameGregorian
+            case StringFormatter.CalendarJulian => CalendarNameJulian
+            case _ => errorFun
+        }
+    }
+}
+
+/**
+  * Represents the name of a Gregorian or Julian calendar.
+  */
+sealed trait CalendarNameGregorianOrJulian extends CalendarNameV2
+
+/**
+  * Represents the name of the Gregorian calendar.
+  */
+case object CalendarNameGregorian extends CalendarNameGregorianOrJulian {
+    override def toString: String = StringFormatter.CalendarGregorian
+}
+
+/**
+  * Represents the name of the Julian calendar.
+  */
+case object CalendarNameJulian extends CalendarNameGregorianOrJulian {
+    override def toString: String = StringFormatter.CalendarJulian
+}
+
+/**
+  * Represents a date as values that are suitable for constructing human-readable representations.
+  *
+  * @param calendarName the name of the calendar.
+  * @param year         the date's year.
+  * @param maybeMonth   the date's month, if given.
+  * @param maybeDay     the date's day, if given.
+  * @param maybeEra     the date's era, if the calendar supports it. An era is required in Gregorian and
+  *                     Julian calendars.
+  */
+case class CalendarDateV2(calendarName: CalendarNameV2, year: Int, maybeMonth: Option[Int], maybeDay: Option[Int], maybeEra: Option[DateEraV2]) {
+    if (maybeMonth.isEmpty && maybeDay.isDefined) {
+        throw AssertionException(s"Invalid date: CalendarDateV2($calendarName, $year, $maybeMonth, $maybeDay, $maybeEra)")
+    }
+
+    calendarName match {
+        case _: CalendarNameGregorianOrJulian =>
+            if (maybeEra.isEmpty) {
+                throw AssertionException(s"Era is required in calendar $calendarName")
+            }
+
+        case _ => ()
```

Comment:
In this case, we would have a non supported calendar, right? 

Could you explain why there is a `CalendarNameV2` and a trait `CalendarNameGregorianOrJulian`? 

Is your rationale that you would add additional traits for other calendars and include them in the match case statement?

Reply:
`CalendarNameGregorianOrJulian` is for things like this:

```scala
def getGregorianCalendarChangeDate(calendarName: CalendarNameGregorianOrJulian): Date
```

That method should only accept a Gregorian or Julian calendar, and using the type `CalendarNameGregorianOrJulian` guarantees that this is the case.

Reply:
> In this case, we would have a non supported calendar, right?

That's actually not possible with this code. The `case _` is just there for the future, when we have more calendars. Until then, I could remove that line, because it can never be reached.

The design pattern here is called *sealed case objects*, which is an alternative to using Scala enumerations. See:

https://pedrorijo.com/blog/scala-enums/

With sealed case objects, the compiler can guarantee that the pattern matching is exhaustive.

Reply:
> Is your rationale that you would add additional traits for other calendars and include them in the match case statement?

Yes.

Reply:
We're already using sealed case objects for ontology schemas:

https://github.com/dhlab-basel/Knora/blob/5f7df5d04535463ca9acfbfbd7f36083b3da3981/webapi/src/main/scala/org/knora/webapi/OntologySchema.scala

Reply:
The other advantage of sealed case objects is that you can make a hierarchy of subtypes, like `CalendarNameV2` -> `CalendarNameGregorianOrJulian` -> `CalendarNameGregorian`, which you can't do with enumerations.